File: 0059-spiral-matrix-ii/0059-spiral-matrix-ii.py
class Solution:
    def generateMatrix(self, n):
        matrix = [[0]*n for _ in range(n)]
        num = 1
        top, bottom, left, right = 0, n-1, 0, n-1

        while True:
            for i in range(left, right+1):
                matrix[top][i] = num
                num += 1
            top += 1
            if top > bottom: break

            for i in range(top, bottom+1):
                matrix[i][right] = num
                num += 1
            right -= 1
            if left > right: break

            for i in range(right, left-1, -1):
                matrix[bottom][i] = num
                num += 1
            bottom -= 1
            if top > bottom: break

            for i in range(bottom, top-1, -1):
                matrix[i][left] = num
                num += 1
            left += 1
            if left > right: break

        return matrix

File: 0059-spiral-matrix-ii/0059-spiral-matrix-ii.kt
class Solution {
    fun generateMatrix(n: Int): Array<IntArray> {
        val matrix = Array(n) { IntArray(n) }
        var top = 0
        var bottom = n - 1
        var left = 0
        var right = n - 1
        var num = 1

        while (true) {
            for (i in left..right) matrix[top][i] = num++
            if (++top > bottom) break

            for (i in top..bottom) matrix[i][right] = num++
            if (--right < left) break

            for (i in right downTo left) matrix[bottom][i] = num++
            if (--bottom < top) break

            for (i in bottom downTo top) matrix[i][left] = num++
            if (++left > right) break
        }
        return matrix
    }
}
